File: src/com/anathema_roguelike/entities/characters/player/perks/abilities/spells/SpellSpecializationChoice.java
package com.anathema_roguelike.entities.characters.player.perks.abilities.spells;

import com.anathema_roguelike.entities.characters.Character;
import com.anathema_roguelike.entities.characters.player.classes.PlayerClass;
import com.anathema_roguelike.entities.characters.player.perks.abilities.Ability;
import com.anathema_roguelike.entities.characters.player.perks.specializations.AbilitySpecializationChoice;

public class SpellSpecializationChoice extends AbilitySpecializationChoice {
	
	private int spellLevel;
	private Class<? extends PlayerClass> casterClass;
	
	public SpellSpecializationChoice(int spellLevel, Class<? extends PlayerClass> casterClass) {
		super(Spell.class);
		
		this.spellLevel = spellLevel;
		this.casterClass = casterClass;
	}
	
	@Override
	public boolean validAbility(Character character, Ability a) {
		return super.validAbility(character, a) && ((Spell<?>)a).getCasterClass() == casterClass && ((Spell<?>)a).getSpellLevel() == spellLevel;
	}
}

File: src/com/anathema_roguelike/entities/characters/player/perks/abilities/spells/SpellSpecializationChoice.scala
package com.anathema_roguelike
package entities.characters.player.perks.abilities.spells

import com.anathema_roguelike.entities.characters.player.classes.PlayerClass
import com.anathema_roguelike.entities.characters.player.perks.abilities.Ability
import com.anathema_roguelike.entities.characters.player.perks.specializations.AbilitySpecializationChoice
import com.anathema_roguelike.entities.characters.Character

class SpellSpecializationChoice(var spellLevel: Int, var casterClass: Class[_ <: PlayerClass])
  extends AbilitySpecializationChoice[Spell[_]] {

  override def validAbility(character: Character, a: Ability): Boolean = {
    super.validAbility(character, a) && (a match {
      case spell: Spell[_] => spell.getCasterClass == casterClass && spell.getSpellLevel == spellLevel
      case _ => false
    })
  }
}